ui/src/app/views/dataset/component.tsx

```diff
@@ -169,13 +196,53 @@ export const DataSet = withCurrentWorkspace()(class extends React.Component<
     navigateByUrl(`/workspaces/${namespace}/${id}/cohorts/${cohort.id}/review`);
   }
 
+  getDomainsFromConceptIds(selectedConceptSetIds: number[]): Domain[] {
+    const {conceptSetList} = this.state;
+    return fp.uniq(conceptSetList.filter((conceptSet: ConceptSet) =>
+      selectedConceptSetIds.includes(conceptSet.id))
+      .map((conceptSet: ConceptSet) => conceptSet.domain));
+  }
+
+  async getValuesList(domains: Domain[]): Promise<ValueSet[]> {
+    const {namespace, id} = this.props.workspace;
+    const valueSets = fp.zipWith((domain: Domain, valueSet: DomainValuesResponse) =>
+        ({domain: domain, values: valueSet.items}),
+      domains,
+      await Promise.all(domains.map((domain) =>
+        conceptsApi().getValuesFromDomain(namespace, id, domain.toString()))));
+    return valueSets;
+  }
+
   select(resource: ConceptSet | Cohort, rtype: ResourceType): void {
     if (rtype === ResourceType.CONCEPT_SET) {
+      const {valueSets, selectedValues} = this.state;
       const origSelected = this.state.selectedConceptSetIds;
+      let newSelectedConceptSets: number[];
       if (fp.includes(resource.id, origSelected)) {
-        this.setState({selectedConceptSetIds: fp.remove((c) => c === resource.id, origSelected)});
+        newSelectedConceptSets = fp.remove((c) => c === resource.id, origSelected) as number[];
       } else {
-        this.setState({selectedConceptSetIds: (origSelected).concat(resource.id)});
+        newSelectedConceptSets = (origSelected).concat(resource.id);
+      }
+      const currentDomains = this.getDomainsFromConceptIds(newSelectedConceptSets);
+      const origDomains = valueSets.map(valueSet => valueSet.domain);
+      const newDomains = fp.without(origDomains, currentDomains) as unknown as Domain[];
+      const removedDomains = fp.without(currentDomains, origDomains);
+      const updatedValueSets =
+        valueSets.filter(valueSet => !(fp.contains(valueSet.domain, removedDomains)));
+      const updatedSelectedValues =
+        selectedValues.filter(selectedValue =>
+          !fp.contains(selectedValue.domain, removedDomains));
+      if (newDomains.length > 0) {
+        this.getValuesList(newDomains)
+          .then(newValueSets => this.setState((state) => ({
```

Comment:
If we're not using this `state`, then using the callback form here is unnecessary.